docs: Add integrated storage concepts page (#8673)
hashicorp/vault: website/pages/docs/concepts/integrated-storage.mdx

integrated storage.

## Cluster Membership

This section will outline how to bootstrap and manage a cluster of Vault nodes
running integrated storage. 

Integrated storage is bootstrapped during the [initialization
process](https://learn.hashicorp.com/vault/getting-started/deploy#initializing-the-vault),
and results in a cluster of size 1. Depending on the [desired deployment
size](/docs/internals/integrated-storage/#deployment-table), nodes can be joined
to the active Vault node.

### Joining Nodes

Joining is the process of taking an uninitialized Vault node and making it a
member of an existing cluster. In order to authenticate the new node to the
cluster it must use the same seal mechanism. If using a Auto Unseal the node
must be configured to use the same KMS provider and Key as the cluster it's
attempting to join. If using a Shamir seal the unseal keys must be provided to
the new node before the join process can complete. Once a node has successfully
joined, data from the active node can begin to replicate to it. Once a node has
been joined it cannot be re-joined to a different cluster.

You can either join the node automatically via the config file or manually through the

> API. When joining a node, the API address of the leader node must be used. We

Reviewer: ```suggestion
API (both methods described below). When joining a node, the API address of the leader node must be used. We
```

---
layout: docs
page_title: Integrated Storage
sidebar_title: Integrated Storage
description: Learn about the integrated raft storage in Vault.
---

# Integrated Storage

> Vault supports a number of Storage options for the durable storage of Vault's

Reviewer 2: ```suggestion
Vault supports a number of storage options for the durable storage of Vault's
```

information. As of Vault 1.4 an integrated storage option is offered. This
storage backend does not rely on any third party systems, it implements high
availability, supports Enterprise Replication features, and provides
backup/restore work flows.

The integrated storage option stores Vault's data on the server's filesystem and
uses a consensus protocol to replicate data to each server in the cluster. More
information on the internals of integrated storage can be found in the
[integrated storage internals
documentation](/docs/internals/integrated-storage/). Also the 
[Configuration](/docs/configuration/storage/raft/) docs can help in configuring

> Vault to used integrated storage.

Reviewer 2: ```suggestion
Vault to use integrated storage.
```

The sections below go into various details on how to operate Vault with
integrated storage.

## Cluster Membership

This section will outline how to bootstrap and manage a cluster of Vault nodes
running integrated storage. 

Integrated storage is bootstrapped during the [initialization
process](https://learn.hashicorp.com/vault/getting-started/deploy#initializing-the-vault),
and results in a cluster of size 1. Depending on the [desired deployment
size](/docs/internals/integrated-storage/#deployment-table), nodes can be joined
to the active Vault node.

### Joining Nodes

Joining is the process of taking an uninitialized Vault node and making it a
member of an existing cluster. In order to authenticate the new node to the
cluster it must use the same seal mechanism. If using a Auto Unseal the node
must be configured to use the same KMS provider and Key as the cluster it's
attempting to join. If using a Shamir seal the unseal keys must be provided to
the new node before the join process can complete. Once a node has successfully

> joined data from the active node can begin to replicate to it. Once a node has

Reviewer 2: ```suggestion
joined, data from the active node can begin to replicate to it. Once a node has
```

been joined it cannot be re-joined to a different cluster.

> You can either join the node automatically via the config file or manually through the

Reviewer 2: Link to the respective config here would be helpful.

Reviewer: I thought that the first time I read it too, but the configs are described right below this, so I wasn't sure how useful it would be.

> API. When joining a node the API address for the leader node must be used. We

Reviewer 2: ```suggestion
API. When joining a node, the API address for the leader node must be used. We
```

node3    node3.vault.local:8201    leader      true
```

## Server-to-Server Communication

Once nodes are joined to one another they begin to communicate using mTLS over
Vault's cluster port. The cluster port defaults to `8201`. The TLS information
is exchanged at join time and is rotated on a cadence. 

A requirement for integrated storage is that the
[cluster_addr](/docs/concepts/ha/#per-node-cluster-address) configuration option
is set. This allows Vault to assign an address to the node ID at join time.

## Outage Recovery

### Quorum Maintained

This section outlines the steps to take when a single server or multiple servers
are in a failed state but quorum is still maintained. This means the remaining
alive servers are still operational, can elect a leader, and are able to process
write requests. 

If the failed server is recoverable, the best option is to bring it back online
and have it reconnect to the cluster with the same host address. This will return
the cluster to a fully healthy state. 

If this is impractical you need to remove the failed server. Usually, you can
issue a remove-peer command to remove the failed server if it's still a member
of the cluster.

> If the remove-peer command isn't possible or you'd rather manually re-write the

Reviewer 2: We may want to link this section to the peers.json section just the the next section is doing using `See the section below on manual recovery using peers.json for details of the
recovery procedure`.

> cluster membership a `raft/peers.json` file can be written to the configured

Reviewer 3: We may want to mention whether this file can be removed (or still needs to exist) once quorum has be re-established.

Author: Vault will delete it automatically.

data directory.

### Quorum Lost

In the event that multiple servers are lost, causing a loss of quorum and a

> complete outage partial recovery is still possible.

Reviewer: ```suggestion
complete outage, partial recovery is still possible.
```

If the failed servers are recoverable, the best option is to bring them back
online and have them reconnect to the cluster using the same host addresses.
This will return the cluster to a fully healthy state. 

> If the failed servers are not recoveryable partial recovery is possible using

Reviewer: ```suggestion
If the failed servers are not recoverable, partial recovery is possible using
```

data on the remaining servers in the cluster. There may be data loss in this
situation because multiple servers were lost, so information about what's
committed could be incomplete. The recovery process implicitly commits all
outstanding Raft log entries, so it's also possible to commit data that was
uncommitted before the failure.

See the section below on manual recovery using peers.json for details of the

> recovery procedure. You simply include just the remaining servers in the

Reviewer: ```suggestion
recovery procedure. You include only the remaining servers in the
```

raft/peers.json recovery file. The cluster should be able to elect a leader once
the remaining servers are all restarted with an identical raft/peers.json
configuration.

> Any new servers you introduce later can be fresh with totally clean data

Reviewer: ```suggestion
Any servers you introduce later can be fresh with totally clean data
```

directories and joined using Vault's join command.

In extreme cases, it should be possible to recover with just a single remaining
server by starting that single server with itself as the only peer in the
raft/peers.json recovery file.

### Manual Recovery Using peers.json

Using raft/peers.json for recovery can cause uncommitted Raft log entries to be
implicitly committed, so this should only be used after an outage where no other

> option is available to recover a lost server. Make sure you don't have any

Reviewer: Maybe bold the "Make sure you don't have"

The next step is to go to the [configured data
path](/docs/configuration/storage/raft/#path) of each Vault server. Inside that
directory, there will be a raft/ sub-directory. We need to create a
raft/peers.json file. The file should be formatted as a JSON array containing
the node ID, address:port, and suffrage information of each Vault server you
wish to be in the cluster.

```json
[
  {
    "id": "node1",
    "address": "node1.vault.local:8201",
    "non_voter": false
  },
  {
    "id": "node2",
    "address": "node2.vault.local:8201",
    "non_voter": false
  },
  {
    "id": "node3",
    "address": "node3.vault.local:8201",
    "non_voter": false
  }
]
```

* id (string: <required>) - Specifies the node ID of the server. This can be
  found in the logs when the server starts up if it was auto-generated, and it
  can also be found inside the node-id file in the server's data directory.
* address (string: <required>) - Specifies the host and port of the server. The
  port is the server's cluster port.
* non_voter (bool: <false>) - This controls whether the server is a non-voter.
  If omitted, it will default to false, which is typical for most clusters.

> Simply create entries for all servers. You must confirm that servers you do not

Reviewer: ```suggestion
Create entries for all servers. You must confirm that servers you do not
```

Using raft/peers.json for recovery can cause uncommitted Raft log entries to be
implicitly committed, so this should only be used after an outage where no other
option is available to recover a lost server. Make sure you don't have any
automated processes that will put the peers file in place on a periodic basis.

To begin, stop all remaining servers. 

The next step is to go to the [configured data
path](/docs/configuration/storage/raft/#path) of each Vault server. Inside that
directory, there will be a raft/ sub-directory. We need to create a
raft/peers.json file. The file should be formatted as a JSON array containing
the node ID, address:port, and suffrage information of each Vault server you
wish to be in the cluster.

```json
[
  {
    "id": "node1",
    "address": "node1.vault.local:8201",
    "non_voter": false
  },
  {
    "id": "node2",
    "address": "node2.vault.local:8201",
    "non_voter": false
  },
  {
    "id": "node3",
    "address": "node3.vault.local:8201",
    "non_voter": false
  }
]
```

* id (string: <required>) - Specifies the node ID of the server. This can be
  found in the logs when the server starts up if it was auto-generated, and it

>   can also be found inside the node-id file in the server's data directory.

Reviewer 2: Or in the config file that was used to bring up the Vault node?

* address (string: <required>) - Specifies the host and port of the server. The
  port is the server's cluster port.
* non_voter (bool: <false>) - This controls whether the server is a non-voter.

>   If omitted, it will default to false, which is typical for most clusters.

Reviewer 2: Do we want to mention that this is an enterprise feature here?
